tools/expotools/src/commands/ClientBuild.ts

```diff
@@ -1,144 +1,177 @@
 import { Command } from '@expo/commander';
-import spawnAsync from '@expo/spawn-async';
-import { Config, UpdateVersions } from '@expo/xdl';
 import aws from 'aws-sdk';
 import chalk from 'chalk';
 import fs from 'fs-extra';
 import inquirer from 'inquirer';
 import path from 'path';
 
-import { STAGING_API_HOST, EXPO_DIR, IOS_DIR, ANDROID_DIR } from '../Constants';
-import { Platform, iosAppVersionAsync, androidAppVersionAsync } from '../ProjectVersions';
+import { EXPO_DIR } from '../Constants';
 import askForPlatformAsync from '../utils/askForPlatformAsync';
+import askForSdkVersionAsync from '../utils/askForSDKVersionAsync';
+import { link } from '../Formatter';
 import Git from '../Git';
+import logger from '../Logger';
+import { modifySdkVersionsAsync, getSdkVersionsAsync } from '../Versions';
+import { ClientBuilder, Platform } from '../client-build/types';
+import IosClientBuilder from '../client-build/IosClientBuilder';
+import AndroidClientBuilder from '../client-build/AndroidClientBuilder';
+
+const s3Client = new aws.S3({ region: 'us-east-1' });
+const { yellow, blue, magenta } = chalk;
 
 type ActionOptions = {
-  platform?: 'ios';
+  platform?: Platform;
   release: boolean;
+  skipUpload: boolean;
+};
+
+export default (program: Command) => {
+  program
+    .command('client-build')
+    .alias('cb')
+    .description(
+      'Builds Expo client for iOS simulator or APK for Android, uploads the archive to S3 and saves its url to versions endpoint.'
+    )
+    .option('-p, --platform [string]', 'Platform for which the client will be built.')
+    .option(
+      '-r, --release',
+      'Whether to upload and release the client build to staging versions endpoint.',
+      false
+    )
+    .option(
+      '--skip-upload',
+      "Whether to skip the uploading part. Might be useful for debugging and not to unintentionally override production's build.",
+      false
+    )
+    .asyncAction(main);
 };
 
-async function askToRecreateSimulatorBuildAsync(archivePath: string) {
+async function main(options: ActionOptions) {
+  const platform = options.platform || (await askForPlatformAsync());
+  const sdkBranchVersion = (await Git.getSDKVersionFromBranchNameAsync()) || '39.0.0';
+
+  if (options.release && !sdkBranchVersion) {
+    throw new Error(`Client builds can be released only from the release branch!`);
+  }
+
+  const builder = getBuilderForPlatform(platform);
+  const sdkVersion = sdkBranchVersion || (await askForSdkVersionAsync(platform));
+  const appVersion = await builder.getAppVersionAsync();
+
+  await buildOrUseCacheAsync(builder);
```

Comment:
Why not to change it to sth like this?:
```suggestion
  await builder.buildOrUseCacheAsync();
```
This seem as the more natural way of handling this kind of operation.

Reply:
Well, I don't think it should be a part of the builder — it wraps some builder methods, but it also has some additional logic (for caching) that are not much related to the distinction of the builders.